### ISSUE TITLE
issue when multiple threads accessing the registeredCalls dictionary

### ISSUE THREAD

**Issue by vigdora** [danielsaidi/Mockery#20]
when multiple threads are trying to access mock.registeredCalls it crashes.
any suggestion to a thread safe solution?

**Comment by danielsaidi** [comment]
Hi @vigdora 

I haven't run into this, do you use multithreading in your tests?

**Comment by vigdora** [comment]
Hi @danielsaidi 
when I run tests using XCTestCase, I see that multiple threads are running. 
upon executing the next lines inside `registerCall`, it crashes:
``` 
let calls = mock.registeredCalls[ref.id] ?? [] 
mock.registeredCalls[ref.id] = calls + [call]
```
          
  when I wrapped this code with serialQueue, it stopped crashing:
  
  ```   
  serialQueue.sync {
            let calls = mock.registeredCalls[ref.id] ?? []
            mock.registeredCalls[ref.id] = calls + [call]
        }
 ```
 do you run unitests on one thread only? 
 tbh, I am not sure how to force  XCTestCase to run only on one thread,
 and even so, I suspect it might impact Performance.

**Comment by danielsaidi** [comment]
I usually don't change the default Xcode test setup, and assumed that running tests in parallel would still use isolated processes. So if you see this crash in the default setup, some form of thread safety is probably needed.

**Comment by vigdora** [comment]
@danielsaidi thanks for the quick reply.
it doesn"t happen when running tests in parallel. it actually happen upon running singular test suite. (with multiple tests inside it)
for example: 

```
func fetchData() async throws {
   async let dataX = try await oneRepo.fetchX()
   async let dataY = await oneRepo.fetchY()
       
        let _: [Any] = try await [dataX, dataY ]
}
```
now lets assume oneRepo is mocked, 
upon running the test, I have noticed the fetchX, and fetchY are running on different threads - 
thats where the crash happening. 
ofcourse I have simplified the code for the sake of example (:

**Comment by danielsaidi** [comment]
Thank you, that's great information 👍 

Async functions run on different threads, so that is most definitely the problem. Perhaps it would be worth making the mock a main actor?

**Comment by vigdora** [comment]
yee I think thats a possible solution  ( I will make sure it works in the meantime...)
but do you think it worth achieving an allistic solution,  such that developers wont need to worry about it when they create a mock?
it took me some time to investigate (;
but anyWay, @mainActor might work here, as you suggested
and regardless, your package is really nice solution over all

**Comment by danielsaidi** [comment]
I most definitively want to add such a solution within the library itself, so devs don't have to do anything.

I only wish I can find a way to reproduce the problem.

**Comment by vigdora** [comment]
@danielsaidi 
so I tried adding @MainActor on the mocked class and indeed the mocked methods now run only on the main thread.
**but it didn't solve the issue**, since inside async registerCall function I still see **multiple thread running**. (probably calling to `call` function is changing the thread as well)
you can try to reproduce the issue by creating something similar to the example I showed before, 
and than run the test **repeatedly around 100-200 times**. 
if it still won't reproduce I will create a branch for you with the issue later this week - let me know

**Comment by danielsaidi** [comment]
@vigdora 

Thank you for continuing to look at this! I am unfortunately very busy with a few other projects, so I will not have time to look at this problem right now. 

If you find a way to solve it, I'd be happy to merge it and release a new version. 👍

**Comment by tboogh** [comment]
We had similar issues with this so in our fork we use a semaphore to guard the access, it's not the most elegant solution but as a triage it works. https://github.com/BookBeat/MockingKit/tree/thread-safe

**Comment by danielsaidi** [comment]
@vigdora and @tboogh I haven't looked into this in a while, since I don't use the library that often these days. I enabled Swift 6 and strict concurrency without getting any warnings or errors. This was a bit surprising, since we may have these threading issue, but I'm not sure how to best approach it.

**Comment by tboogh** [comment]
@danielsaidi If you add `Sendable` conformance to `Mock` you'll inform the compiler that it should perform strict checking on it and it will fail to compile with the Swift 6 language mode.

**Comment by danielsaidi** [comment]
@tboogh Huh, that's strange. I was under the impression that enabling Swift 6 for the package would automatically catch any potential threading issues.

I'm not sure how to even start working on this, since the whole point of the `Mock` class is to be open to inheritance.

**Comment by tboogh** [comment]
Yeah, I completely understand the current implementations strengths and the intention of the design. I’m testing different approaches but haven’t landed in one that will be Swift 6 friendly yet either.On 17 Mar 2025, at 09:02, Daniel Saidi ***@***.***> wrote:﻿
  @tboogh Huh, that's strange. I was under the impression that enabling Swift 6 for the package would automatically catch any potential threading issues.
I'm not sure how to even start working on this, since the whole point of the Mock class is to be open to inheritance.—Reply to this email directly, view it on GitHub, or unsubscribe.You are receiving this because you were mentioned.Message ID: ***@***.***>



danielsaidi left a comment (danielsaidi/MockingKit#20)
@tboogh Huh, that's strange. I was under the impression that enabling Swift 6 for the package would automatically catch any potential threading issues.
I'm not sure how to even start working on this, since the whole point of the Mock class is to be open to inheritance.

**Comment by danielsaidi** [comment]
@tboogh @vigdora I've been giving this some thought. When are you facing the situation where the same mock is used in multiple threads? 

Do you by any chance use a singleton in your test case, like this:

```swift
import MockingKit
import XCTest

final class GenericTests: XCTestCase {

    let mock = MyMock()

    ....
}
```

or do you set it up like this?

```swift
import MockingKit
import XCTest

final class GenericTests: XCTestCase {

    let mock: MyMock!

    override func setUp() {
        mock = MyMock()
    }
}
```

The latter case should work just fine, since a new mock is then created for each run. Or does it still happen in this case?

**Comment by Enricoza** [comment]
Sorry to jump in, 
I'm not actually a user of the library (yet) but was looking at it and evaluating if I could use it when I found this issue.

In the last message you sent, @danielsaidi, I think you are missing the fact that when running the tests for a `XCTestCase` class, that class gets instanciated once **per each test case**. 
As a result to this, you can see that your two approaches are actually the same: `MyMock` instance will not be reused among different test cases. 

You can verify what I'm saying by having 2 test cases like so:
```swift
final class GenericTests: XCTestCase {

    let mock = MyMock()

    func test1() {
        print(mock)
    }

    func test2() {
        print(mock)
    }
}
```
And verify that the two prints should point out to two different objects with different memory addresses.

On the other hand, if you do something like this:
```swift
import MockingKit
import XCTest

let mock = MyMock()

final class GenericTests: XCTestCase {
    func test1() {
        print(mock)
    }

    func test2() {
        print(mock)
    }
}
```
The mock will be shared and the memory addresses will be the same.

I doubt, anyway, that people are actually creating shared Mocks in this way.

---

I'm just gonna put here a wild guess, maybe it's useful. 
I think the threading issue here lies within the single test case: multiple accesses to those registered calls from different threads. 
I'm not sure if this is due to the way the library works internally (haven't dug deeply in the source code yet) or if it's a usage issue (where the users are switching the threads and the library doesn't support that), but I've had similar crashes on some of my tests that use the same logic as your library does (but with a lot more boilerplate on each mock).

Solution for me was to always use non thread-safe classes from a single thread only. Not sure how that would look in your case though.

**Comment by danielsaidi** [comment]
Hi @Enricoza 

Thank you, I thought that the class was created once for all tests, and the setup function was then used to adjust the same class instance for each test case.

I haven't experienced the threading issues myself, but I had hoped that adding support for Swift 6 concurrency would help pinpointing the source, but the root cause seems to elude the type system.

**Comment by Enricoza** [comment]
Yes, @danielsaidi, I thought the same for many years, as most developers do, since the API seems to suggest that.

Unfortunately the Swift 6 concurrency doesn't solve those types of issues.

If you want and have time to debug this, I would suggest to try the [thread sanitizer](https://developer.apple.com/documentation/xcode/diagnosing-memory-thread-and-crash-issues-early#Detect-data-races-among-your-apps-threads) if you haven't yet. 

It works wonders and allows you to spot race conditions at runtime even if those race conditions are not met during that specific run, enhancing significantly your ability to reproduce the issue.

**Comment by danielsaidi** [comment]
Thank you @Enricoza 🙏

**Comment by danielsaidi** [comment]
I think @tboogh's sendable and lock changes will fix this. If not, feel free to reopen.
